packages/classes/src/PayloadArray.js

```diff
@@ -0,0 +1,100 @@
+import Payload from './Payload'
+
+/**
+ * PayloadArray stores a series of Payloads in a data structure that allows for easy filtering
+ * and lookup by position and type.
+ */
+class PayloadArray {
+	payloads = []
+
+	constructor(...payloads) {
+		[...payloads].forEach((payload) => this.insert(payload))
+	}
+
+	get length() { return this.payloads.length }
+
+	/**
+	 * Insert a Payload into the PayloadArray while maintaining position order.
+	 *
+	 * @param  {Payload} payload The Payload to be inserted.
+	 * @return {PayloadArray}    This PayloadArray to allow for chaining.
+	 */
+	insert = (payload) => {
+		if (!(payload instanceof Payload)) {
+			throw new Error('Attempt to insert a non-Payload to PayloadArray')
+		}
+		const index = this.indexOfPosition(payload.position)
+		this.payloads.splice(index, 0, payload)
+		return this
+	}
+
+	/**
+	 * Find the index of the first Payload whose position is greater than or equal to this one.
+	 *
+	 * If no qualifying payloads exist, this will return the length of the PayloadArray, because
+	 * the implied position for the payload is the end of the array.
+	 *
+	 * @param  {Number} position The position being searched for.
+	 * @return {Number}          The index of the item, or the PayloadArray length.
+	 */
+	indexOfPosition = (position) => {
+		const recursiveSearch = (sortedPayloads, target, left, right) => {
+			if (left === right) return left
+			const index = Math.floor((left + right) / 2)
+			if (index === sortedPayloads.length) return index
+			const payload = sortedPayloads[index]
+			if (payload.position === target) return index
+			if (payload.position < target) {
+				return recursiveSearch(sortedPayloads, target, index + 1, right)
+			}
+			if (payload.position > target) return recursiveSearch(sortedPayloads, target, left, index)
+			return index
+		}
+		return recursiveSearch(this.payloads, position, 0, this.payloads.length)
+	}
+
+	/**
+	 * Filter out the Payloads of the specified type.
+	 *
+	 * NOTE: This returns a new PayloadArray and does not modify the original.
+	 *
+	 * @param  {String} type  The type of payload to include in the new PayloadArray.
+	 * @return {PayloadArray} A filtered PayloadArray.
+	 */
+	filterByType = (type) => {
+		const filteredPayloadArray = new PayloadArray()
+		filteredPayloadArray.payloads = this.payloads.filter(
+			(payload) => RegExp(`^${type.replace(/\.ANY$/, '.*')}$`).test(payload.type),
```

Comment:
May be more efficient to create the RegExp outside the filter loop.

Reply:
Also may want to escape the regexp if you're letting outside values in, or validate the values. Probably this is supposed to be a clean value in a closed system, so not actually vulnerable, but my guess is static analysis would now flag this as susceptible to a known DoS attack ([ReDoS](https://nodegoat.herokuapp.com/tutorial/redos)) that has caused a wave of NPM "vulnerability" warnings.

Reply:
Oh TIL about that -- I think we can just drop regex entirely here and use .startsWith

Reply:
That would probably be faster and safer.